### PR TITLE
Support function handlers

### PR DESCRIPTION
**Pull request by akshay-ap**
This PR is for merging into main branch.
Fixes: #47

Includes all changes of #52

Changes in PR:

- Create FunctionHandlerManager contract
- Store function handler per Safe per function id
- Tests
- Test return data value from handle function
- Rename function modifier onlyPermittedPlugin to onlyPermittedIntegration and appropriate event, function parameter names
- All functions that update state of a Safe e.g. `enablePlugin`, `disablePlugin`, `setHooks`, `setFunctionHandler` have `onlySelf` function modifier.

### CODE REVIEW

File: contracts/base/OnlySelfCallable.sol [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    function checkOnlySelf() private view {
+        if (msg.data.length < 20) {
+            revert InvalidCalldataLength();
+        }
+        address sender;
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            sender := shr(96, calldataload(sub(calldatasize(), 20)))
+        }
+        if (sender != msg.sender) {
```
Comment (akshay-ap) [inline contracts/base/OnlySelfCallable.sol:22]: too strict check here?

Comment (akshay-ap) [reply contracts/base/OnlySelfCallable.sol:22]: This is will block relayed transactions

Comment (rmeissner) [reply contracts/base/OnlySelfCallable.sol:22]: Hmm why would it block this? The relaying would happen on Safe level not Manager level, right?

File: contracts/base/OnlySelfCallable.sol [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+abstract contract OnlySelfCallable {
+    error InvalidSender(address sender);
+    error InvalidCalldataLength();
+
+    modifier onlySelf() {
+        checkOnlySelf();
+        _;
+    }
+
+    function checkOnlySelf() private view {
```
Comment (rmeissner) [inline contracts/base/OnlySelfCallable.sol:13]: I find this name very misleading, as `address(this)` has no impact on this check.

File: contracts/base/OnlySelfCallable.sol (hunk 1 above)
Comment (akshay-ap) [reply contracts/base/OnlySelfCallable.sol:22]: Yes, if transaction is relayed to Safe and then Manager is called then no issues.

File: contracts/base/OnlySelfCallable.sol (hunk 2 above)
Comment (akshay-ap) [reply contracts/base/OnlySelfCallable.sol:13]: then `checkCallerisSender()`?

File: contracts/base/OnlySelfCallable.sol [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.18;
+
+abstract contract OnlySelfCallable {
+    error InvalidSender(address sender);
+    error InvalidCalldataLength();
+
+    modifier onlySelf() {
```
Comment (rmeissner) [inline contracts/base/OnlySelfCallable.sol:8]: Similar to my comment below (which was resolved) this function doesn't make sense for me. Same the contract name. "Self" here is the manager contract, which is not relevant in any check. We are checking that the call comes from the account ... so something like "onlyAccount" or similar would make more sense

File: contracts/base/OnlySelfCallable.sol [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        checkCallerisSender();
+        _;
+    }
+
+    function checkCallerisSender() private view {
```
Comment (rmeissner) [inline contracts/base/OnlySelfCallable.sol:13]: This method is severely lacking comments. We use assembly and do length checks, both should be well commented and if possible linked to standards this is based on.

File: contracts/base/RegistryManager.sol [hunk 5]
```diff
@@ -29,8 +25,22 @@ contract RegistryManager is Ownable2Step {
...
+        // Only allow registered and non-flagged integrations
+        (uint64 listedAt, uint64 flaggedAt) = ISafeProtocolRegistry(registry).check(integration);
+        if (listedAt == 0 || flaggedAt != 0) {
+            revert IntegrationNotPermitted(integration, listedAt, flaggedAt);
+        }
+    }
+
     /**
      * @notice Allows only owner to update the address of a registry. Emits event RegistryChanged(egistry, newRegistry)
+     * @dev Do not set owner as a Safe address with Manager enabled as fallback handler.
```
Comment (rmeissner) [inline contracts/base/RegistryManager.sol:43]: Could we add a "why" here

File: test/FunctionHandlerManager.spec.ts [hunk 6]
```diff
@@ -0,0 +1,193 @@
+import { SignerWithAddress } from "@nomicfoundation/hardhat-ethers/signers";
+import hre, { deployments, ethers } from "hardhat";
+import { getMockFunctionHandler } from "./utils/mockFunctionHandlerBuilder";
+import { IntegrationType } from "./utils/constants";
+import { expect } from "chai";
+import { getInstance } from "./utils/contracts";
+import { MaxUint256, ZeroAddress } from "ethers";
+import { ISafeProtocolFunctionHandler__factory, MockContract } from "../typechain-types";
+
+describe("Test Function Handler", async () => {
```
Comment (rmeissner) [inline test/FunctionHandlerManager.spec.ts:10]: What do we normally put in the top level `describe`? I would have expected the contract name.

**Review by rmeissner — CHANGES_REQUESTED** [review]

File: contracts/base/RegistryManager.sol (hunk 5 above)
Comment (akshay-ap) [reply contracts/base/RegistryManager.sol]: Updated and Moved it to constructor: https://github.com/5afe/safe-core-protocol/pull/70/commits/7acc9cc13058551a41eb78f877f78fc35c9e9a6f

File: contracts/base/OnlySelfCallable.sol (hunk 2 above)
Comment (akshay-ap) [reply contracts/base/OnlySelfCallable.sol]: updated to `onlyAccount`

File: contracts/base/RegistryManager.sol [hunk 7]
```diff
@@ -9,18 +9,22 @@ contract RegistryManager is Ownable2Step {
...
-            revert PluginNotPermitted(plugin, listedAt, flaggedAt);
-        }
+    modifier onlyPermittedIntegration(address integration) {
+        checkPermittedIntegration(integration);
         _;
     }
 
+    /*
+     * @notice Constructor that sets registry address and owner address.
+     * @dev Do not set owner as a Safe address with Manager enabled as fallback handler.
```
Comment (akshay-ap) [inline contracts/base/RegistryManager.sol:22]: Will be tackled in https://github.com/5afe/safe-core-protocol/issues/75

**Pull request merged**